### ISSUE TITLE
Memory leak with dynamicScenes if sharedMemoryForWorkers is turned off

### ISSUE THREAD

**Issue by saschandroid** [mkkellogg/GaussianSplats3D#129]
Hi!

I wanted to turn off `sharedMemoryForWorkers` for my project/tests but if I use the viewer configuration

`'sharedMemoryForWorkers': false`

together with

`'dynamicScenes': true`

the memory consumption of the chrome browser is increasing about 100mb per sec. This happens with each of the demo scenes if I use these settings (e.g. the garden scene or the dynamic scene). Is this the expected behaviour and for `dynamicScenes` we always have to use `sharedMemoryForWorkers` or is this a bug?

**Comment by mkkellogg** [comment]
For what it's worth, I am not seeing a memory leak under those same conditions in Chrome. I took `demo/dynamic_scenes.html` and added `'sharedMemoryForWorkers': false` to the viewer config and Chrome's JS heap stayed steady around 95 Mb. Can you share your exact source code, maybe I can reproduce that way. Also, what version of Chrome are you using?

**Comment by saschandroid** [comment]
I checked out the source code this morning and did not change anything except for adding `'sharedMemoryForWorker': false` in dynamic_scenes.html and `'sharedMemoryForWorker': false, 'dynamicScenes': true` in garden.html. When opening the taskmanager I can see that chrome is using around 8GB of memory after a few seconds and is still consuming more and more with no other tabs or chrome windows open. The scenes where I did not add the `'sharedMemoryForWorker': false, 'dynamicScenes': true`  combination are working fine.

In short I did the following:
I took the source code from this mornings checkout, added the two lines and did `npm install`, `npm run build-windows` and `npm run demo`,  then started chrome with `http://127.0.0.1:8080/index.html`. 

I'm on Windows 10 with Chrome 121.0.6167.185

**Comment by mkkellogg** [comment]
Just verifying -- You passed `sharedMemoryForWorker` and `dynamicScenes` as parameters to the `Viewer` constructor?  The actual parameter names are `sharedMemoryForWorkers` and `dynamicScene`.

**Comment by saschandroid** [comment]
Ups, sorry typo. This is the code:
```
const viewer = new GaussianSplats3D.Viewer({
      'cameraUp': [0, -1, -0.54],
      'initialCameraPosition': [-7.14242, -2.25111, -2.36519],
      'initialCameraLookAt': [1.52976, 2.27776, 1.65898],
      'dynamicScene': true,
      'sharedMemoryForWorkers': false
    });
```

**Comment by mkkellogg** [comment]
I tried on a Windows 11 machine with the latest version of Chrome (121.0.6167.185) and I still could not reproduce the issue. Does it happen in Firefox for you as well?

**Comment by mkkellogg** [comment]
So I was finally able to reproduce the issue in the latest version of Chrome in Windows 11. For some reason, Chrome is just not performing garbage collection on the sort results that are returned from the web worker and that's why it's eating up all the memory. If I open up the "Memory" tab and repeatedly force garbage collection by pressing the garbage can icon, the issue doesn't occur.

My recommendation for now would be to not use dynamic scene mode if you're going to disable shared memory. Even with garbage collection working, there's going to be a big performance hit because of all the memory cleanup that will need to be done. I will probably need to add some documentation about this to the README.

**Comment by mkkellogg** [comment]
Actually I have a fix now that might technically reduce performance just a tad, but should resolve the memory leak issue. Can you check out the `web_worker_memory_leak` branch and see if that works for you? Thanks!

**Comment by saschandroid** [comment]
Thanks for your work!
With the fix the memory leak is gone. I had it also in Edge and Opera; Firefox used a lot of memory but the GC was working.

I was testing the code on a mobile device, so I turned off gpu and shared memory with dynamic scene enabled (I'm trying to move/rotate the splats in AR). That's how I found out about the memory leak. Another reason for not using shared memory were the potential security issues. I'm not a web developer and mainly working on 3d stuff like reconstruction, 3d file formats, visualization, etc., so I tried the 'easy' way by just disabling shared memory. 

And yes, I used dynamic scene with shared memory enabled because of the performance ... with reporting the issue I was just trying to be helpful ;-) Again: thank you for fixing the issue so quickly

**Comment by mkkellogg** [comment]
Glad to hear that it worked! Thank you for finding this issue, I don't think I would have ever noticed it myself :) I merged the fix into `main`, hopefully I'll get a new release out with it soon.
